# Nota de Estudo — Compressão em Dois Estágios com *Intecooler* (Solução de Sistema Não-Linear)

> **Autor:** Fábio Magnani (fabio.magnani@ufpe.br)  
> **Curso:** Engenharia Mecânica - UFPE  
> **Início do desenvolvimento:** 29/09/2025  
> **Primeira publicação:** 06/10/2025   
> **Versão Atual:** v-2025-10-06-a (fase de teste técnico no Colab, teste didático-pedagógico com estudantes e início da revisão final de código/texto)

---

## Objetivo
Fazer a modelagem matemática e obter a solução de um processo de compressão em dois estágios com *intercooler*, Prob. 6-15 do Stoecker.

## Referência
- Stoecker, W. F. *Design of Thermal Systems*, chap. 6, 3rd ed., McGraw-Hill, 1989.

## Estrutura

- Introdução
- Modelo matemático
- Solução
- Estudo paramétrico (trabalhos futuros)
- Propriedades termodinâmicas reais (trabalhos futuros)
- Redução do número de equações (trabalhos futuros)
- Conclusão

## Notação

(em construção)

## Parâmetros

(em construção)

## Hipóteses simplificativas
- Regime permanente  
- Gás perfeito e líquido incompressível com calor específico constante  
- Equação do trocador na forma exponencial modificada  



## Revisão

Em geral, temos à disposição três tipos de equação para usar:    
    - Leis de conservação (*e.g.,*: conservação da massa, conservação da energia)     
    - Relações entre as propriedades termodinâmicas    
    - Características dos equipamentos (*e.g.*, eficiência conhecida, alguma propriedade constante)     

<figure style="margin: 0 0 1.25rem 0; text-align:center;">
  <img src="../figures/ne-compressao-dois-estagios-intercooler-fig-01.png"
       alt=""
       width="600" style="display:block; margin:0 auto; max-width:100%;">
  <figcaption><b>Figura 1. Compressão de ar em dois estágios com resfriamento intermediário (*intercooler*).</b>  (Prob. 6-15 do Stoecker).</figcaption>
</figure> 

## Enunciado do Problema

Um compressor de dois estágios com *intercooler* (Fig.01, Prob 6-15 do Stoecker) comprime ar seco de 100 kPa absolutos para 1200 kPa absolutos. Os seguintes dados valem para os componentes:

- **Taxa de deslocamento volumétrico (vazão teórica)**    
    -- Estágio baixo: 0.2 m³/s    
    -- Estágio alto: 0.05 m³/s         
      
- **Eficiência volumétrica** (para cada compressor): $\eta_v = (\dot V_{\text{real}}/\dot V_{\text{desl}})$, 
onde $\dot V_{\text{real}}$ é a vazão volumétrica medida na sucção e $\dot V_{\text{desl}}$ é a taxa de deslocamento volumétrico (vazão teórica).

- **Para ambos os compressores**, a eficiência é: $\eta = 1,04 - 0,04\,\left(p_{\text{desc}}/p_{\text{suc}}\right)^{1.4}$.

- **Expoente politrópico**: O compressor de baixa pressão respeita a equação $p_1\,v_1^{n} = p_2\,v_2^{n}$, 
com $n =  1.2$

- **Intercooler**: trocador de calor de contra-corrente, recebendo água a 22°C com vazão de 0,09 kg/s; o coeficiente do trocador de calor é $UA = 0.3\ \text{kW/K}$.

- **Hipótese**: o ar comporta-se como gás perfeito com $c_{p0}$ constante, a água comporta-se como líquido incompressível com $c$ constante, e o trocador de calor é isobárico.

## Objetivo    

Determinar as grandezas do sistema (por exemplo $w$, $p_i$, $T_2$, $T_3$) de acordo com os dados acima.



## Parâmetros

$\dot V_{D,\text{baixa}}=0,20\ \text{m}^3/\text{s}$  
$\dot V_{D,\text{alta}}=0,05\ \text{m}^3/\text{s}$  
$n_{\text{baixa}}=1{,}2$  
$UA=0,3\ \text{kW/K}$  
$p_1=100\ \text{kPa}$  
$T_1=26^\circ\text{C}$  
$p_4=1200\ \text{kPa}$  
$w_w=0,09\ \text{kg/s}$  
$T_{w,e}=22^ \circ\text{C}$  
$R=0,286\ \text{kJ/kg} \cdot \text{K}$  
$c_{p,a}=1,0\ \text{kJ/kg} \cdot \text{K}$   
$c_{w}=4,22\ \text{kJ/kg} \cdot \text{K}$

---

## Incógnitas

- $w,\ \dot Q$  
- $v_1,\ v_2,\ v_3$  
- $p_i$  
- $T_2,\ T_3,\ T_{w,o}$

---

## Equações

1. (Eficiência vol. comp. baixa) $w=\dfrac{V_{D,\text{baixa}}}{v_1}\,\Big[1,04-0,04 \left( \frac{p_i}{p_1} \right) ^{1,4}\Big]$
2. (Gás perfeito) $p_1\,v_1=R\,T_1$
3. (Equação do compressor de baixa) $p_1\,v_1^{n_{\text{baixa}}}=p_i\,v_2^{n_{\text{baixa}}}$
4. (Gás perfeito) $p_i\,v_2=R\,T_2$
5. (1ª lei, lado quente do trocador, ar) $\dot Q=w c_{p,a}\,(T_2-T_3)$
6. (1ª lei, lado frio do trocador, água) $\dot Q=w_w c_{w}\,(T_{w,s}-T_{w,e})$
7. (Equação do trocador, LMTD) $(T_{2}-T_{w,s}) =  (T_{3}-T_{w,e}) \exp \left[ \frac{(T_{2}-T_{w,s})-(T_{3}-T_{w,e})}{Q/UA} \right]$
8. (Eficiência vol. comp. baixa) $w=\dfrac{V_{D,\text{alta}}}{v_3}\,\Big[1,04-0,04 \left( \frac{p_4}{p_i} \right) ^{1,4}\Big]$
9. (Gás perfeito) $p_i\,v_3=R\,T_3$


---

**Observação**    

Para eliminar problemas de logarítmo de número negativo e divisão por zero na equação tradicional 
do trocador, na 7a equação do sistema, trocamos a forma   
$$
\dot Q = UA \frac{(T_{qe}-T_{fs})-(T_{qs}-T_{fe})}{\ln(\frac{(T_{qe}-T_{fs})}{(T_{qs}-T_{fe})})}
$$
por
$$
(T_{qe}-T_{fs}) =  (T_{qs}-T_{fe}) \exp \left[ \frac{(T_{qe}-T_{fs})-(T_{qs}-T_{fe})}{\dot Q/UA} \right]
$$


In [3]:
import numpy as np  # NumPy para operações numéricas

# -------------------------
# Parâmetros (unidades SI)
# -------------------------
Vd_baixa = 0.20      # m^3/s  # Vazão volumétrica efetiva do 1º estágio (baixa pressão)
Vd_alta  = 0.05      # m^3/s  # Vazão volumétrica efetiva do 2º estágio (alta pressão)
n_baixa  = 1.2       # [-]    # Expoente politrópico do 1º estágio
UA       = 0.3       # kW/K   # Condutância global do intercooler (UA)
p1       = 100.0     # kPa    # Pressão de sucção do 1º estágio
T1C      = 26.0      # °C     # Temperatura de entrada do ar (em °C)
p4       = 1200.0    # kPa    # Pressão de descarga final
w_w      = 0.09      # kg/s   # Vazão mássica de água no intercooler (dado)
Tw_e_C   = 22.0      # °C     # Temperatura de entrada da água (em °C)
R        = 0.286     # kJ/(kg·K) # Constante específica do ar (coerente com cp e Q)
cp_a     = 1.0       # kJ/(kg·K) # Calor específico do ar (suposição)
c_w      = 4.22      # kJ/(kg·K) # Calor específico da água (suposição)

# Temperaturas em K
T1   = T1C + 273.15  # K  # Conversão da temperatura de entrada do ar para Kelvin
Tw_e = Tw_e_C + 273.15  # K  # Conversão da temperatura de entrada da água para Kelvin

# Chutes iniciais
p_i0 = (p1+p4)/2                 # kPa  # Pressão intermediária entre os estágios (palpite)
T2_0 = 500                       # K    # Temperatura após o 1º estágio (palpite)
T3_0 = (T2_0+Tw_e)/2             # K    # Temperatura após o intercooler (entre T2_0 e Tw_e)
v1_0 = R * T1/p1                  # m^3/kg # Volume específico na sucção do 1º estágio (gás ideal)
v2_0 = R * T2_0 / p_i0            # m^3/kg # Volume específico no ponto intermediário
v3_0 = R * T3_0 / p4              # m^3/kg # Volume específico na descarga final
Q_0  = 60       # kW              # Carga térmica no intercooler (palpite)
w_0  = Q_0/(T2_0-T3_0)            # kg/s  # Vazão de ar coerente com Q_0 por balanço no ar
Tw_s0 = Tw_e + Q_0/0.4            # K     # Saída da água estimada (palpite com coeficiente efetivo)

x0 = np.array([w_0, Q_0, v1_0, v2_0, v3_0, p_i0, T2_0, T3_0, Tw_s0], dtype=float)  # Vetor inicial das incógnitas

# limites inferiores e superiores
lb = np.array([0, 1,   0, 0, 0, 0,  T1,   T1,   0], dtype=float)           # Limites inferiores (fisicalidade)
ub = np.array([1, 500, 2, 2, 2, p4, 1000, 1000, 1000], dtype=float)        # Limites superiores (faixas plausíveis)

# -------------------------
# Vetor de incógnitas:
# x = [w, Q, v1, v2, v3, p_i, T2, T3, Tw_s]
# -------------------------
def residual(x):  # Função de resíduos para least_squares
    w, Q, v1, v2, v3, p_i, T2, T3, Tw_s = x  # Desempacotamento das variáveis
    
    r1 = w - (Vd_baixa / v1) * (1.04 - 0.04 * (p_i / p1)**1.4)  # Balanço volumétrico/mássico no 1º estágio (correção empírica)
    r2 = p1 * v1 - R * T1                                       # Gás ideal na sucção: p1*v1 = R*T1
    r3 = p1 * (v1**n_baixa) - p_i * (v2**n_baixa)               # Relação politrópica no 1º estágio: p*v^n = const
    r4 = p_i * v2 - R * T2                                      # Gás ideal no ponto intermediário: p_i*v2 = R*T2
    r5 = Q - w * cp_a * (T2 - T3)                               # Balanço de energia no ar: Q = w*cp*(T2−T3)
    r6 = Q - w_w * c_w * (Tw_s - Tw_e)                          # Balanço de energia na água: Q = w_w*c_w*(Tw_s−Tw_e)
    dT1 = (T2 - Tw_s)                                           # ΔT1 para LMTD: ar quente de entrada vs água de saída
    dT2 = (T3 - Tw_e)                                           # ΔT2 para LMTD: ar de saída vs água de entrada
    r7 = dT1  - dT2*np.exp(UA/Q*(dT1 - dT2))                   # Modelo do trocador (LMTD): Q = UA*(ΔT1−ΔT2)/ln(ΔT1/ΔT2)
    r8 = w - (Vd_alta / v3) * (1.04 - 0.04 * (p4 / p_i)**1.4)   # Balanço volumétrico/mássico no 2º estágio (correção empírica)
    r9 = p_i * v3 - R * T3                                      # Gás ideal na descarga: p_i*v3 = R*T3

    return np.array([r1, r2, r3, r4, r5, r6, r7, r8, r9], dtype=float)  # Vetor de resíduos (9 equações)


In [5]:
from scipy.optimize import least_squares  # Solver de mínimos quadrados com suporte a bounds

def solve_box(fun, x0, lb, ub):  # Resolve min 0.5*||r(x)||^2 com limites (caixa)
    res = least_squares(fun, x0, bounds=(lb, ub), method="trf", xtol=1e-9, ftol=1e-9, gtol=1e-9, max_nfev=2000)  # TRF com tolerâncias rígidas
    r = fun(res.x)  # Recalcula resíduos no ponto ótimo encontrado
    print("status =", res.status, "|", res.message)  # Mensagem do solver (convergência/motivo de parada)
    print("nfev =", res.nfev, "| ||res||_2 =", np.linalg.norm(r))  # Avaliações e norma-2 dos resíduos
    return res.x, r  # Retorna solução e vetor de resíduos

x_star_b, res_b = solve_box(residual, x0, lb, ub)  # Chamada com função de resíduos, chute e bounds
print("x*_box =", x_star_b)  # Vetor solução (w, Q, v1, v2, v3, p_i, T2, T3, Tw_s)
print("res_box =", res_b)    # Resíduos nas 9 equações no ponto solucionado


status = 1 | `gtol` termination condition is satisfied.
nfev = 181 | ||res||_2 = 2.6167399688845543e-14
x*_box = [1.80770093e-01 1.04685662e+01 8.55569000e-01 2.76571292e-01
 2.33855435e-01 3.87737254e+02 3.74954522e+02 3.17043581e+02
 3.22713365e+02]
res_box = [-2.77555756e-17  0.00000000e+00 -1.42108547e-14 -1.42108547e-14
 -7.10542736e-15  5.32907052e-15  0.00000000e+00  2.77555756e-17
  1.42108547e-14]


## Discussão dos Resultados

**Condições resolvidas (SI):**
- Vazão mássica de ar `w` ≈ **0,1808 kg/s**
- Calor removido no intercooler `Q` ≈ **10,47 kW**
- Pressão intermediária `p_i` ≈ **388 kPa**
- Temperaturas do ar:  
  `T2` ≈ **375 K** (≈ 102 °C), `T3` ≈ **317 K** (≈ 44 °C)  
- Água: `Tw_e` = **295,15 K** (22 °C), `Tw_s` ≈ **322,7 K** (≈ 49,6 °C)
- Volumes específicos: `v1` ≈ **0,8556 m³/kg**, `v2` ≈ **0,2766 m³/kg**, `v3` ≈ **0,2339 m³/kg**

**Checks de consistência (todas atendidas a ~1e−14):**
- Estado (gás ideal): `p1 v1 = R T1`, `p_i v2 = R T2`, `p_i v3 = R T3`
- Balanço de energia do ar: `Q = w · c_p (T2 − T3)` → **10,47 kW**
- Balanço no lado da água: `Q = w_w c_w (Tw_s − Tw_e)` → **10,47 kW**
- LMTD: ΔT₁ = `T2 − Tw_s` ≈ **52,24 K**, ΔT₂ = `T3 − Tw_e` ≈ **21,89 K**  
  `ΔT_lm = (ΔT₁ − ΔT₂)/ln(ΔT₁/ΔT₂)` ≈ **34,90 K**  
  `UA·ΔT_lm` = **0,3 × 34,90 ≈ 10,47 kW = Q**
- Vazões volumétricas (com eficiência volumétrica empírica):  
  η_v,1 = `1,04 − 0,04 (p_i/p1)^1,4` ≈ **0,773**  
  η_v,2 = `1,04 − 0,04 (p4/p_i)^1,4` ≈ **0,845**  
  `w = (Vd_baixa/v1) η_v,1 = (Vd_alta/v3) η_v,2` → **0,1808 kg/s** em ambos os estágios

**Interpretação física:**
- **p_i acima da média geométrica.** A média geométrica `√(p1 p4)` ≈ **346 kPa**. O resultado **388 kPa** desloca a razão para cima porque os deslocamentos são **diferentes** (`Vd_baixa = 0,20` vs `Vd_alta = 0,05 m³/s`). Para compatibilizar as vazões, o modelo eleva `p_i` (aumenta a densidade na sucção do 2º estágio), o que reduz `v3` e permite que o compressor de **menor deslocamento** entregue a mesma `w`.
- **Intercooler efetivo, mas limitado por UA.**  
  Capacidades térmicas:  
  lado quente `C_hot = w c_p ≈ 0,181 kW/K`, lado frio `C_cold = w_w c_w ≈ 0,380 kW/K`.  
  Assim, `C_min = C_hot` e `NTU = UA/C_min ≈ 1,66`, `C_r = C_min/C_max ≈ 0,476`.  
  **Efetividade** `ε = Q / (C_min (T2 − Tw_e)) ≈ 0,726`.  
  Isso explica por que `T3` (≈ 44 °C) **não** cai até `Tw_e` (22 °C): o limite é o produto `UA`.
- **Temperaturas e aproximações:**  
  Aproximação no frio (pinch frio) `T3 − Tw_e` ≈ **21,9 K**; no quente `T2 − Tw_s` ≈ **52,2 K**. Estes valores são coerentes com `NTU` moderado.

**Leituras úteis dos resultados:**
- Aumentar **UA** ou **w_w** (mantendo `UA` efetivo maior) reduziria `T3` e `Tw_s`, elevando `Q`.
- Se os deslocamentos fossem iguais, `p_i` tenderia à **média geométrica**; com deslocamentos distintos, o `p_i` ótimo se desloca para **balancear vazões**.
- As curvas de eficiência volumétrica penalizam razões de compressão elevadas; isso também empurra `p_i` para um valor que equalize `w` em ambos os estágios.

**Conclusão:**  
O conjunto está **fechado e fisicamente consistente**: estados ideais satisfeitos, balanços de energia nos dois lados, LMTD igualando `UA·ΔT_lm` a `Q` e compatibilidade de vazões via deslocamentos e eficiências volumétricas. O intercooler é razoavelmente eficaz (ε ≈ 0,73), mas `UA` limita a aproximação à temperatura de entrada da água.


### Trabalhos Futuros

1. **Estudo paramétrico**
   - Variar parâmetros de projeto para compreender sensibilidade e desempenho do sistema:
     - UA (efetividade térmica do intercooler)
     - Vazões volumétricas e razões de deslocamento (Vd_baixa, Vd_alta)
     - Vazões mássicas de ar e água (w, w_w)
     - Expoente politrópico e pressões intermediárias
   - Automatizar a geração de gráficos (T2, T3, Q, p_i) em função desses parâmetros.

2. **Propriedades termodinâmicas reais**
   - Substituir o gás ideal pelo **ar real** via **CoolProp** (funções `PropsSI`).
   - Incluir correções de `cp(T)`, viscosidade e densidade em função da temperatura.
   - Preparar versão compatível com **Google Colab**, incluindo:
     - Instalação automática do CoolProp (`!pip install CoolProp`)
     - Verificação de versão e exemplo de uso rápido (`PropsSI('D','T',300,'P',1e5,'Air')`)

3. **Redução do número de equações**
   - Eliminar redundâncias:
     - Usar relações ideais para substituir variáveis intermediárias (`v = RT/p`).
     - Integrar balanço de energia diretamente nas equações principais.
   - Reduzir o sistema para **5–6 incógnitas** (w, Q, p_i, T2, T3, Tw_s) mantendo fechamento físico.
   - Reescrever o modelo em forma vetorial compacta, com normalização das grandezas.

---

**Objetivo:** evoluir o modelo didático atual em direção a uma ferramenta paramétrica, realista e leve, pronta para uso interativo no *Colab* e integrada às futuras atividades do *EspiralLab*.
